pyphare/pyphare/core/operators.py

```diff
@@ -0,0 +1,185 @@
+        )
+        ds_z[nb_ghosts:-nb_ghosts, nb_ghosts:-nb_ghosts].fill(
+            0.0
+        )  # TODO at 2D, gradient is null in z dir
+
+    else:
+        raise RuntimeError("dimension not yet implemented")
+
+    return (
+        {"name": "x", "data": ds_x, "centering": patch_data["scalar"].centerings},
+        {"name": "y", "data": ds_y, "centering": patch_data["scalar"].centerings},
+        {"name": "z", "data": ds_z, "centering": patch_data["scalar"].centerings},
+    )
+
+
+def dot(hier_left, hier_right, **kwargs):
+    names_left_kept = hier_left.get_names()
+    names_right_kept = hier_right.get_names()
+
+    if isinstance(hier_left, VectorField) and isinstance(hier_right, VectorField):
+        names_left = ["left_x", "left_y", "left_z"]
+        names_right = ["right_x", "right_y", "right_z"]
+
+    else:
+        raise RuntimeError("type of hierarchy not yet considered")
+
+    hl = rename(hier_left, names_left)
+    hr = rename(hier_right, names_right)
+
+    h = compute_hier_from(
+        _compute_dot_product,
+        (hl, hr),
+    )
+
+    # hier_left = rename(hl, names_left_kept)
+    # hier_right = rename(hr, names_right_kept)
+
+    return ScalarField(
+        h.patch_levels,
+        h.domain_box,
+        refinement_ratio=h.refinement_ratio,
+        time=h.times()[0],
+        data_files=h.data_files,
+    )
+
+
+def cross(hier_left, hier_right, **kwargs):
+    names_left_kept = hier_left.get_names()
+    names_right_kept = hier_right.get_names()
+
+    if isinstance(hier_left, VectorField) and isinstance(hier_right, VectorField):
+        names_left = ["left_x", "left_y", "left_z"]
+        names_right = ["right_x", "right_y", "right_z"]
+
+    else:
+        raise RuntimeError("type of hierarchy not yet considered")
+
+    hl = rename(hier_left, names_left)
+    hr = rename(hier_right, names_right)
+
+    h = compute_hier_from(
+        _compute_cross_product,
+        (hl, hr),
+    )
+
+    # hier_left = rename(hl, names_left_kept)
+    # hier_right = rename(hr, names_right_kept)
+
+    return VectorField(
+        h.patch_levels,
+        h.domain_box,
+        refinement_ratio=h.refinement_ratio,
+        time=h.times()[0],
+        data_files=h.data_files,
+    )
+
+
+def sqrt(hier, **kwargs):
+    h = compute_hier_from(
+        _compute_sqrt,
+        hier,
+    )
+
+    return ScalarField(
+        h.patch_levels,
+        h.domain_box,
+        refinement_ratio=h.refinement_ratio,
+        time=h.times()[0],
+        data_files=h.data_files,
+    )
+
+
+def modulus(hier):
+    assert isinstance(hier, VectorField)
+
+    return sqrt(dot(hier, hier))
+
+
+def grad(hier, **kwargs):
+    assert isinstance(hier, ScalarField)
+    nb_ghosts = list(hier.level(0).patches[0].patch_datas.values())[0].ghosts_nbr[0]
+    h = compute_hier_from(_compute_grad, hier, nb_ghosts=nb_ghosts)
+
+    # TODO the plot of a grad displays only 1 patch if vmin and vmax are not specified... any idea why ?
```

Comment:
I see you told rabbit this is fixed, should we thus rm the comment?

Reply:
yep, done